contract/models/contract_line.py

```diff
@@ -0,0 +1,819 @@
+                    raise ValidationError(
+                        _(
+                            "Contract '%s' start date can't be later than "
+                            "end date"
+                        )
+                        % line.contract_id.name
+                    )
+
+    @api.depends('recurring_next_date', 'date_start', 'date_end')
+    def _compute_create_invoice_visibility(self):
+        today = fields.Date.today()
+        for line in self:
+            if line.date_start:
+                if today < line.date_start:
+                    line.create_invoice_visibility = False
+                elif not line.date_end:
+                    line.create_invoice_visibility = True
+                elif line.recurring_next_date:
+                    if line.recurring_invoicing_type == 'pre-paid':
+                        line.create_invoice_visibility = (
+                            line.recurring_next_date <= line.date_end
+                        )
+                    else:
+                        line.create_invoice_visibility = (
+                            line.recurring_next_date
+                            - line.get_relative_delta(
+                                line.recurring_rule_type,
+                                line.recurring_interval,
+                            )
+                        ) <= line.date_end
+
+    @api.model
+    def recurring_create_invoice(self, contract=False):
+        domain = []
+        date_ref = fields.Date.today()
+        if contract:
+            contract.ensure_one()
+            date_ref = contract.recurring_next_date
+            domain.append(('contract_id', '=', contract.id))
+
+        domain.extend(
+            [
+                ('contract_id.recurring_invoices', '=', True),
+                ('recurring_next_date', '<=', date_ref),
+                ('is_canceled', '=', False),
+                # '|',
+                # ('date_end', '=', False),
+                # ('date_end', '>=', date_ref),
+                # with this leaf, it's impossible to invoice the last period
+                # in post-paid case.
+                # i.e: date_end = 15/03 recurring_next_date = 31/03
+                # A solution for this, is to only check recurring_next_date
+                # and filter with create_invoice_visibility
+            ]
+        )
+        lines = self.search(domain).filtered('create_invoice_visibility')
+        if lines:
+            return lines._recurring_create_invoice()
+        return False
+
+    @api.multi
+    def _recurring_create_invoice(self):
+        """Create invoices from contracts
+
+        :return: invoices created
+        """
+        invoices = self.env['account.invoice']
+        for contract in self.mapped('contract_id'):
+            lines = self.filtered(lambda l: l.contract_id == contract)
+            invoices |= lines._create_invoice()
+            lines._update_recurring_next_date()
+        return invoices
```

Comment:
I'm not sure this method will be efficient if we are invoicing many lines.

Reply:
I changed it with read_group